packages/ui/src/components/Designer/index.tsx

```diff
@@ -112,13 +120,29 @@ const TemplateEditor = ({
             (plugin) => plugin?.propPanel.defaultSchema.type === value
           )?.propPanel;
           Object.assign(tgt, propPanel?.defaultSchema || {});
+        } else if (key === 'position.x' || key === 'position.y') {
+          const padding = isBlankPdf(template.basePdf) ? template.basePdf.padding : [0, 0, 0, 0];
+          const [paddingTop, paddingRight, paddingBottom, paddingLeft] = padding;
+          const { width: pageWidth, height: pageHeight } = pageSizes[pageCursor];
+          const { width: targetWidth, height: targetHeight } = tgt;
+          if (key === 'position.x') {
+            tgt.position.x = Math.min(
+              Math.max(Number(value), paddingLeft),
+              pageWidth - targetWidth - paddingRight
+            );
+          } else {
+            tgt.position.y = Math.min(
+              Math.max(Number(value), paddingTop),
+              pageHeight - targetHeight - paddingBottom
+            );
+          }
```

Comment:
it might be worth extracting this out to a testable function, this is within a really important callback function

Reply:
You're right. This function is becoming complex. I have created a separate issue.

https://github.com/pdfme/pdfme/issues/402

Reply:
ref: https://github.com/pdfme/pdfme/pull/403